pkg/datapath/linux/routing/routing.go

```diff
@@ -118,12 +122,13 @@ func (info *RoutingInfo) Configure(ip net.IP, mtu int, masq bool) error {
 
 // Delete removes the ingress and egress rules that control traffic for
 // endpoints. Note that the routes within these rules are not deleted as they
-// can be reused when another endpoint is created on the same node. The reason
-// for this is that ENI devices under-the-hood are simply network interfaces
-// and all network interfaces have an ifindex. This index is then used as the
-// table ID when these rules are created. The routes are created inside a table
-// with this ID, and because this table ID equals the ENI ifindex, it's stable
-// to rely on and therefore can be reused.
+// can be reused when another endpoint is created on the same node.
+//
+// Note, even though we know egress rule resides in a per-ENI table, we don't
+// know the table ID at delete-time. However, the ingress rule is always in the
+// main routing table. The egress rule is a best-effort search, but it is
+// created under a specific priority, which is unlikely to have collisions with
+// other rules on the system, or so we hope.
```

Comment:
The *rule* in the routing rules list points to a table, where the *routes* live right? I think the existing comment is a little confusing given "routes within these rules" but my understanding is that logically there's a linked list of rules (with priority) which then point into some kind of map/array of tables that decide how to forward traffic.

The ingress/egress distinction as far as I can tell is purely a Cilium construct where we dedicate ingress priority in this initial linked list as 20 and egress priority as 110 , then there's a few other stock standard rules before/in-between/after, including the final priority 32766-32767 rules at the end which point to the main/default routing tables.

I tried to reword a couple of times but got a bit confused about the main point here, I think we're trying to say something like:

> Note that one or more IPs may share the same route table, as identified by the interface number of the corresponding device. This function only removes the egress rules to disconnect the per-ENI egress routes from a specific local IP, and does not remove the corresponding route table as other IPs may still be using that table.

And maybe,

> The search for the egress rule corresponding to this IP is a best-effort based on the egress priority that Cilium uses, which we assume full control over. That is, if any other component in the system uses the same priority, this function would delete all such rules as well.

(Well I don't know whether the "delete all such rules" part is correct, but having a known consequence or even worst case just "behaviour is undefined if this happens" might be nice detail to point out here.)

Reply:
> The rule in the routing rules list points to a table, where the routes live right? I think the existing comment is a little confusing given "routes within these rules" but my understanding is that logically there's a linked list of rules (with priority) which then point into some kind of map/array of tables that decide how to forward traffic.

Yes you're correct and that was what I was intending to express by saying "within" to denote some sort of linkage between the two. Your wording is much clearer and more concise so I'll borrow it :). Modifying it to mention ingress rules as well.

> (Well I don't know whether the "delete all such rules" part is correct, but having a known consequence or even worst case just "behaviour is undefined if this happens" might be nice detail to point out here.)

The deletion matches on the IP and the priority. If we find more than one rule matching, we skip deletion altogether. So it's not accurate to say that we will delete all rules with a certain priority.